### PR TITLE
Documentation improvements for classes in Objective Function Pieces

### PR DESCRIPTION
**Pull request by ghwilliams**
@dccowan Please take a look at this tentative improvement of the documentation. 

#### Summary

Add proper docstrings to classes and methods in `data_misfit.py` and `objective_function.py`.

#### PR Checklist
* [x] If this is a work in progress PR, set as a Draft PR
* [x] Linted my code according to the [style guides](https://docs.simpeg.xyz/content/getting_started/contributing/code-style.html).
* [x] Added [tests](https://docs.simpeg.xyz/content/getting_started/practices.html#testing) to verify changes to the code.
* [x] Added necessary documentation to any new functions/classes following the
      expect [style](https://docs.simpeg.xyz/content/getting_started/practices.html#documentation).
* [x] Marked as ready for review (if this is was a draft PR), and converted 
      to a Pull Request
* [x] Tagged ``@simpeg/simpeg-developers`` when ready for review.

### CODE REVIEW

**Comment by ghwilliams** [comment]
I've just noticed that I missed some corrections, doing it right now.

**Comment by ghwilliams** [comment]
I've finished what I think was necessary for this PR to be ready for a review.

**Comment by dccowan** [comment]
@ghwilliams please add summary and keep PR checklist up to date. Suggest:

**Summary:** Add proper docstrings for data_misfit.py and objective_function.py (and simulation.py?)

File: docs/conf.py [hunk 1]
```diff
@@ -45,7 +45,7 @@
     "sphinx.ext.extlinks",
     "sphinx.ext.intersphinx",
     "sphinx.ext.mathjax",
-    "sphinx_gallery.gen_gallery",
+    #    "sphinx_gallery.gen_gallery",
```
Comment (dccowan) [inline docs/conf.py:48]: Don't forget to uncomment

File: SimPEG/data_misfit.py [hunk 2]
```diff
@@ -8,12 +8,10 @@
 
 
 class BaseDataMisfit(L2ObjectiveFunction):
```
Comment (dccowan) [inline SimPEG/data_misfit.py:10]: Base classes are a good place to add descriptive docstring for what the .py file actually has. And we need to define the input parameters.

File: SimPEG/data_misfit.py [hunk 3]
```diff
@@ -95,23 +119,44 @@ def nP(self):
...
     @property
     def shape(self):
-        """"""
+        """Shape of the Jacobian.
+
+        The number of data by the number of model parameters.
+
+        Returns
+        -------
+        list of int (n_data, n_param)
```
Comment (santisoler) [inline SimPEG/data_misfit.py:139]: Just a small correction on the type of the returned object:
```suggestion
        tuple of int (n_data, n_param)
```

File: SimPEG/data_misfit.py [hunk 4]
```diff
@@ -68,11 +87,12 @@ def debug(self, value):
 
     @property
     def counter(self):
-        """Set this to a ``SimPEG.utils.Counter`` if you want to count things.
+        """SimPEG ``Counter`` object to store iterations and run-times.
 
         Returns
         -------
-        SimPEG.utils.Counter or None
+        None, SimPEG.utils.Counter
```
Comment (santisoler) [inline SimPEG/data_misfit.py:94]: Small change:
```suggestion
        None or SimPEG.utils.Counter
```

File: SimPEG/data_misfit.py [hunk 5]
```diff
@@ -154,43 +199,91 @@ def W(self, value):
         self._W = value
 
     def residual(self, m, f=None):
+        r"""Computes the data residual vector for a given model.
+
+        Where :math:`\mathbf{d}_obs` is the observed data vector and :math:`\mathbf{d}_{pred}`
```
Comment (santisoler) [inline SimPEG/data_misfit.py:204]: Minor correction in LaTeX equation:
```suggestion
        Where :math:`\mathbf{d}_\text{obs}` is the observed data vector and :math:`\mathbf{d}_\text{pred}`
```

File: SimPEG/data_misfit.py [hunk 6]
```diff
@@ -154,43 +199,91 @@ def W(self, value):
...
+        is the predicted data vector for a model vector :math:`\mathbf{m}`, this function
+        computes the data residual:
+
+        .. math::
+            \mathbf{r} = \mathbf{d}_{pred} - \mathbf{d}_{obs}
```
Comment (santisoler) [inline SimPEG/data_misfit.py:209]:
```suggestion
            \mathbf{r} = \mathbf{d}_\text{pred} - \mathbf{d}_\text{obs}
```

File: SimPEG/data_misfit.py [hunk 7]
```diff
@@ -154,43 +199,91 @@ def W(self, value):
...
+
+        Parameters
+        ----------
+        m : (n_param, ) numpy.ndarray
+            The model for which the function is evaluated.
+        f : None, SimPEG.fields.Fields (optional)
```
Comment (santisoler) [inline SimPEG/data_misfit.py:215]: Small change to follow [numpydocs style for optional parameters](https://numpydoc.readthedocs.io/en/latest/format.html#parameters)
```suggestion
        f : None or SimPEG.fields.Fields, optional
```

File: SimPEG/data_misfit.py [hunk 8]
```diff
@@ -154,43 +199,91 @@ def W(self, value):
...
-    r"""
-    The data misfit with an l_2 norm:
+    r"""Least-squares data misfit.
+
+    Define the data misfit as the L2-norm of the weighted residual between observed
+    data and predicted data for a given model. I.e.:
 
     .. math::
+        \phi_d (\mathbf{m}) = \frac{1}{2} \big \| \mathbf{W_d}
+        \big ( \mathbf{d}_{pred} - \mathbf{d}_{obs} \big ) \big \|_2^2
```
Comment (santisoler) [inline SimPEG/data_misfit.py:237]:
```suggestion
        \big ( \mathbf{d}_\text{pred} - \mathbf{d}_\text{obs} \big ) \big \|_2^2
```

File: SimPEG/data_misfit.py [hunk 9]
```diff
@@ -154,43 +199,91 @@ def W(self, value):
...
+
+    where :math:`\mathbf{d}_obs` is the observed data vector, :math:`\mathbf{d}_{pred}`
```
Comment (santisoler) [inline SimPEG/data_misfit.py:239]:
```suggestion
    where :math:`\mathbf{d}_\text{obs}` is the observed data vector, :math:`\mathbf{d}_\text{pred}`
```

File: SimPEG/data_misfit.py [hunk 10]
```diff
@@ -154,43 +199,91 @@ def W(self, value):
...
+    :math:`\mathbf{W_d}` are the reciprocals of the data uncertainties
+    :math:`\boldsymbol{\varepsilon}`. Thus:
 
-        \mu_\text{data} =
-            \frac{1}{2}
-            \left|
-                \mathbf{W}_d (\mathbf{d}_\text{pred} - \mathbf{d}_\text{obs})
-            \right|_2^2
+    .. math::
+        \mathbf{W_d} = diag \left ( \boldsymbol{\varepsilon}^{-1} \right )
```
Comment (santisoler) [inline SimPEG/data_misfit.py:246]: Minor change in LaTeX style:
```suggestion
        \mathbf{W_d} = \text{diag} \left ( \boldsymbol{\varepsilon}^{-1} \right )
```

File: SimPEG/data_misfit.py [hunk 11]
```diff
@@ -154,43 +199,91 @@ def W(self, value):
...
+
+    Parameters
+    ----------
+    data : SimPEG.data.Data
+        A SimPEG data object that has observed data and uncertainties.
+    simulation : SimPEG.simulation.BaseSimulation
+        A SimPEG simulation object.
+    debug : bool
+        Print debugging information.
+    counter : None, SimPEG.utils.Counter
```
Comment (santisoler) [inline SimPEG/data_misfit.py:256]:
```suggestion
    counter : None or SimPEG.utils.Counter
```

File: SimPEG/data_misfit.py [hunk 12]
```diff
@@ -202,16 +295,32 @@ def deriv(self, m, f=None):
...
-        :param numpy.ndarray v: vector
-        :param SimPEG.fields.Fields f: fields object
+        .. math::
+            \frac{\partial^2 \phi_d}{\partial \mathbf{m}^2} \, \mathbf{v}
+
+        Parameters
+        ----------
+        m : (n_param, ) numpy.ndarray
+            The model for which the Hessian is evaluated.
+        v : None, (n_param, ) numpy.ndarray (optional)
```
Comment (santisoler) [inline SimPEG/data_misfit.py:315]:
```suggestion
        v : None or (n_param, ) numpy.ndarray, optional
```

File: SimPEG/data_misfit.py [hunk 13]
```diff
@@ -202,16 +295,32 @@ def deriv(self, m, f=None):
...
+            A vector.
+
+        Returns
+        -------
+        (n_param, n_param) scipy.sparse.csr_matrix | (n_param, ) numpy.ndarray
```
Comment (santisoler) [inline SimPEG/data_misfit.py:320]:
```suggestion
        (n_param, n_param) scipy.sparse.csr_matrix or (n_param, ) numpy.ndarray
```

File: SimPEG/objective_function.py [hunk 14]
```diff
@@ -11,17 +11,36 @@
...
+    ----------
+    nP : int
+        Number of model parameters.
+    mapping : SimPEG.mapping.BaseMap
+        A SimPEG mapping object that maps from the model space to the
+        quantity evaluated in the objective function.
+    has_fields : bool
+        If ``True``, predicted fields for a simulation and a given model can be
+        used to evaluate the objective function quickly.
+    counter : None, SimPEG.utils.Counter
```
Comment (santisoler) [inline SimPEG/objective_function.py:37]:
```suggestion
    counter : None or SimPEG.utils.Counter
```

File: SimPEG/data_misfit.py [hunk 15]
```diff
@@ -8,12 +8,28 @@
...
+    class inherits the :py:class:`SimPEG.objective_function.L2ObjectiveFunction`.
+    And as a result, it is limited to building data misfit functions of the form:
 
-    .. note::
-        You should inherit from this class to create your own data misfit
-        term.
+    .. math::
+        \phi_d (\mathbf{m}) = \frac{1}{2} \| \mathbf{W} f(\mathbf{m}) \|_2^2
+
+    where :math:`\mathbf{m}` is the model vector, :math:`mathbf{W}` is a linear weighting
```
Comment (santisoler) [inline SimPEG/data_misfit.py:20]: Add missing backslash:
```suggestion
    where :math:`\mathbf{m}` is the model vector, :math:`\mathbf{W}` is a linear weighting
```

File: SimPEG/objective_function.py [hunk 16]
```diff
@@ -104,9 +152,33 @@ def deriv(self, x, **kwargs):
...
+        or the second-derivative (Hessian) multiplied by a vector :math:`(\mathbf{v})`:
+
+        .. math::
+            \frac{\partial^2 \phi}{\partial \mathbf{m}^2} \, \mathbf{v}
+
+        Parameters
+        ----------
+        m : (n_param, ) numpy.ndarray
+            The model for which the Hessian is evaluated.
+        v : None, (n_param, ) numpy.ndarray (optional)
```
Comment (santisoler) [inline SimPEG/objective_function.py:173]:
```suggestion
        v : None or (n_param, ) numpy.ndarray, optional
```

File: SimPEG/objective_function.py [hunk 17]
```diff
@@ -104,9 +152,33 @@ def deriv(self, x, **kwargs):
...
+            A vector.
+
+        Returns
+        -------
+        (n_param, n_param) scipy.sparse.csr_matrix | (n_param, ) numpy.ndarray
```
Comment (santisoler) [inline SimPEG/objective_function.py:178]:
```suggestion
        (n_param, n_param) scipy.sparse.csr_matrix or (n_param, ) numpy.ndarray
```

File: SimPEG/objective_function.py [hunk 18]
```diff
@@ -197,24 +282,26 @@ def __rdiv__(self, denominator):
...
-    multipliers : list or None, optional
-        List containing the multipliers for its respective objective function
+    multipliers : None or list of int, optional
+        List containing the multipliers for each objective function
         in ``objfcts``.  If ``None``, a list full of ones with the same length
         as ``objfcts`` will be created.
-    unpack_on_add : bool, optional
-        Weather to unpack the multiple objective functions when adding them to
+    unpack_on_add : bool
+        Whether to unpack the multiple objective functions when adding them to
```
Comment (santisoler) [inline SimPEG/objective_function.py:304]: Nice catch!

File: SimPEG/objective_function.py [hunk 19]
```diff
@@ -301,24 +389,32 @@ def __getitem__(self, key):
...
-        Multipliers for each objective function
+        r"""Multipliers for the objective functions.
+
+        For a composite objective function :math`\phi` that is a weighted sum of
+        objective functions :math:`\phi_i` with multipliers :math:`c_i` such that
+
+        .. math::
+            \phi = \sum_{i = 1}^N c_i \phi_i
+
+        this method returns the multipliers :math:`c_i` in order.
```
Comment (santisoler) [inline SimPEG/objective_function.py:400]: I think this line could be misleading users to think that multipliers are returned in ascending/descending order. I'm adding this suggestion, let me know what do you think.
```suggestion
        This method returns the multipliers :math:`c_i` in the same order of the ``objfcts``.
```

**Review by santisoler — CHANGES_REQUESTED** [review]
Huge thanks @ghwilliams! Awesome job adding all these docstrings. Users will greatly appreciate it.

I've added a few minor suggestions, mostly corrections to LaTeX equations and some changes to follow numpydocs style.

Feel free to commit the suggestions, or make your own changes if you think they could be improved. Let me know what do you think.

I'm looking forward to merge this and see you in the contributors list! 🚀

**Comment by ghwilliams** [comment]
Thanks @santisoler . I'll review everything and talk to @dccowan to see if we can consider it ready to be merged or if it still needs some corrections.

**Review by santisoler — APPROVED** [review]

**Comment by santisoler** [comment]
Thanks @ghwilliams for applying all the suggestions. I think this is ready to be merged. I've just updated the branch with the latest changes in `main`. We shall wait CI (automated checks and tests) to pass and then we should merge it 🚀 

Thanks again for the contributions! They are greatly appreciated!

**Comment by santisoler** [comment]
Thanks @ghwilliams for all the effort of writing these docstrings! Your contributions are greatly appreciated!

I'm merging this! :rocket:

**Pull request merged**